scijava-types/src/main/java/org/scijava/types/TypeReifier.java

```diff
@@ -110,56 +110,24 @@ default Type reify(final Object o) {
 		}
 
 		final Class<?> c = o.getClass();
-		final TypeVariable<?>[] typeVars = c.getTypeParameters();
-		final int numVars = typeVars.length;
-
-		if (numVars == 0) {
-			// if the class is synthetic, we are probably missing something due to
-			// type erasure.
-			if (c.isSynthetic()) {
-				log.warn("Object " + o + " is synthetic. " +
-					"Its type parameters are not reifiable and thus will likely cause unintended behavior!");
-			}
-			// Object has no generic parameters; we are done!
-			return c;
-		}
-
-		// Object has parameters which need to be resolved. Let's do it.
-
 		// Here we will store all of our object's resolved type variables.
-		final Map<TypeVariable<?>, Type> resolved = new HashMap<>();
-
-		for (final TypeToken<?> token : TypeToken.of(c).getTypes()) {
-			if (resolved.size() == numVars) break; // Got 'em all!
 
-			final Type type = token.getType();
-			if (!Types.containsTypeVars(type)) {
-				// No type variables are buried in this type; it is useless to us!
-				continue;
-			}
-
-			// Populate relevant type variables from the reified supertype!
-			final Map<TypeVariable<?>, Type> vars = //
-				args(o, token.getRawType());
-
-			if (vars != null) {
-				// Remember any resolved type variables.
-				// Note that vars may contain other type variables from other layers
-				// of the generic type hierarchy, which we don't care about here.
-				for (final TypeVariable<?> typeVar : typeVars) {
-					if (vars.containsKey(typeVar)) {
-						resolved.putIfAbsent(typeVar, vars.get(typeVar));
-					}
-				}
+		for (final var token : TypeToken.of(c).getTypes()) {
+			Optional<TypeExtractor> opt = getExtractor(token.getRawType());
+			if (opt.isPresent()) {
+				return opt.get().reify(this, o);
 			}
 		}
 
```

Comment:
Some comments would be appreciated for this new code. They should be at least as witty as my comments that got deleted above. :stuck_out_tongue_closed_eyes:

Reply:
Added in 11cc558e